README.md
Provision-DSL
=============

a simple provisioning toolkit allowing to deploy and configure a remote
machine in a simple way without having to install things on the remote before.

having a config file like:

    {
        # a meaningful name appended to temporary filenames
        name => 'box',
        
        # execute this script on the remote machine
        provision_file => 'box.pl',
        
        # optional definitions for the local machine, eg. minicpan
        local => {
            environment => {
                PERL_CPANM_OPT => "--mirror $ENV{HOME}/minicpan --mirror-only"
            },
        },
        
        # specifications for the remote machine
        remote => {
            hostname => 'box',
        },
        
        # files needed for the provisioning process
        resources => [
            {
                # make everything in 'xxx/resources' except '/dirx'
                #    accessible via Resource('files') during provision
                source      => 'xxx/resources',
                destination => 'files',
                exclude     => 'dirx',
            },
        ],
    }

and a provision file like:

    #!/usr/bin/env perl
    
    use Provision::DSL;
    
    # define variables in another file in the same directory
    # with optional global variables
    include vars, IP => '1.2.3.4', host => 'example.com';
    
    Package 'build-essential';
    # ... more packages
    
    Perlbrew {
        wanted  => '5.16.0',
        install_cpanm => 1,
    };
    
    Dir $WEB_DIR => {
        user => 'root',
        permission => '0755',
    };
    
    Dir $SITE_DIR => {
        user => 'sites',
        content => Resource('website'),
        
        mkdir => [qw(
            logs
            pid
        )],
    };
    
    Done;

by firing the shell command:

    $ provision.pl -c path/to/config.conf

or:

    $ provision.pl -c path/to/config.conf user@host.tld -v

the destination machine will receive the setup from the provision file.
During the provisioning process every entity will check if itself already
is in the right state or leave itself in the desired state.


Changes
Changes for Provision::DSL

0.31    2018-05-11
        * changed a URL to a http: url to fix a failing test

0.30    2018-05-11
        * rsync symlink support (Thanks to Phil Soehnlein)

0.29    2016-03-29
        * Dir entities also allow rsync source as content for retrieving
          their content via rsync instead of locally

0.27    2015-09-24
        * changed default color to terminal's default instead of black
        * added a build script that may run but must exit with code 0
          prior to running the deployment procedure

0.26    2015-07-10
        * improved config file handling

0.25    2015-07-10
        * added possibility to modify config sections via ENV variables
          PROVISION_REMOTE_HOSTNAME
          PROVISION_REMOTE_USER
          PROVISION_REMOTE_ENV_<<env-variable>
          
0.24    2015-07-09
        * improved dryrun handling for entities not executable by defering
          entity instantiation until the latest possible moment
        * added info-logging in combination with -vv switch

0.23    2015-06-23
        * added log_info message for rsync command
        * improved log_info message for dir creation
        
0.22    2013-06-22
        * removed a failing test. TODO: recode this test
        * improved Perl Module checkin of installed modules

0.21    2013-12-02
        * added a PROVISION_HTTP_HOST environment variable allowing to
          specify the host to contact for cpanm module installs from the
          local side.

0.20    2013-08-09
        * added options keyword to Perl_Modules entity to specify additional
          options for cpanm run.

        2013-07-16
        * added Message, Info and Note keywords usable like an entity but
          printing their string message always, under verbose and under
          uncreased verbosity levels. Idea by Stefan Limbacher.
          These messages are optionally controllable by an inspector

        2013-07-15
        * added incremental verbose level
        * added log_info method triggered when verbositiy is higher
        * started adding info-logs to some entities -- more to come.

0.19    2013-07-10
        * added PkgVersion an PodVersion plugins

0.18    2013-07-09
        * increased version for a Pinto test

0.17    2013-07-05
        * share/libs is packed into a .tar file in order to avoid cpan
          servers (like Pinto) to scan the contained packages and get
          confused.

0.16    2013-07-02
        * merge with pre_packed_dists branch. From now on, we have all
          dependencies and the perlbrew install script "on board".
          They are no longer loaded on demand and bundled in versions
          known to work.

0.15    2013-03-27
        * defered calling Source-content to avoid exceptions at entity
          instantiation time
        * re-enabled entity-saving prior to installation because it is
          save now
        * moved install-action from DSL to App in order to allow a dispatcher
          to switch to different possible tasks operating on the same set
          of entities
        * added the concept of tasks allowing a future expansion to filter
          or enable unsually ignored entities grouped by a task name

0.14    2013-03-26
        * updated documentation
        * added Platform::Unix as dependency to explicitly inhibit an
          installation on the forbidden OS
        * changed name for the .provision_... dir to contain host and user
          for keeping distinct directories for individual host/user pairs
        * changed ssh commandline to Net::OpenSSH
        * executed Entities directly at runtime instead of maintaining a list
          to prevent Inspectors from failing too early
        * colorized output of provision run for easier status overview
        * refactored script to multiple Provision::DSL::Local::* classes

        2012-12-12
        * improved restart conditions for Service-Processes under Linux
        * unified clearer/predicate definitions "=> 1" instead of names
        * check if config file returns a hashref, die otherwise

0.13    2012-11-13
        * removed an issue caused by mis-behaved rsync. If 'abc/def' is
          excluded, then 'abc' was reported to get deleted which made no sense
          and caused rsync to periodically fire.
        * removed PERL_CPANM_OPT settings for remote machine

        2012-11-10
        * added docs
        * cleanup Rsync: remove backup
        * Script::Provision cleanup: remove 'Files' keyword
        * made empty provision file handling more tolerant

0.12    2012-11-07
        * introduced File::Zglob for traversal thru directories for
          giving Inspectors simpler parameters for deep file traversal
        * changed the way Inspectors are instantiated
        * Added a base class for inspectors with deep file traversal
        * corrected tests to reflect changes

0.11    2012-11-06
        * added logging to a file inside a log directory if given on the
          commandline.
        * enabling file logging as a default
        * recursive transfer of every log directory content onto the local
          machine after provision.

0.10    2012-11-05
        * simplified the structure of the .provision_XXX directory on the
          local side. Maintaining an identical directory on the remote side.
          Planning to rsync log entries back to local.

0.09    2012-11-04
        * changed the way how inspectors and installers can get used
          in a provision script.
        * updated all tests to match this changed behavior
        * fixed forgotten changes to perl-5.8 compatibility

0.08    2012-11-03
        * removed all absolute paths to executable binaries to be more
          independant from the *nix flavour
        * changed all perl-5.10 specific operations to re-enable things
          when only perl-5.8 is available.

0.07    2012-11-01 (launch date of 007 "skyfall")
        * added "append_if_missing" option to file patch

examples/box.pl
#!/usr/bin/env perl
#
# sample provision file for a host named 'box' (my virtual box)
# will require privileges and may prompt for a password at first launch
#
#     provision.pl -c examples/box.conf
#
use Provision::DSL;
my $WEB_DIR     = '/web/data';
my $SITE_DIR    = "$WEB_DIR/www.mysite.de";

# Package 'build-essential';

# Service 'nginx';

# Perlbrew {
#     # install_cpanm => 1,
#     # wanted  => '5.14.2',
#     wanted  => '5.16.0',
# };

Dir $WEB_DIR => {
    user => 'root',
    permission => '0755',
    # activate to test failing
    # content => '/tmp/xxx/foo'
};

Dir $SITE_DIR => {
    # user => 'vagrant',
    # group => 'vagrant',
    content => Resource('files'),
};

Message 'Completed';    # always visible
Info 'Completed';       # visible in verbose mode
Note 'Completed';       # visible in -vv mode

# die 'stop for testing';

Done;

__END__

User sites => {
    uid => 513,
};

Dir $SITE_DIR;

File "$SITE_DIR/testfile" => {
    content => 'blabla',
};

Perlbrew {
    install_cpanm => 1,
    wanted  => '5.14.2',
};

exit;

Perlbrew sites => {
    install_cpanm => 1,
    wanted        => '5.14.2',
};

File '/path/to/file.ext' => {
    user    => 'sites', # group taken from user
    content => Url('http://domain.tld/path/to/file.ext'),
};

Dir '/path/to/website' => {
    user => 'sites',
    content => Resource('website'), # has a MyApp directory
    
    # MyApp must be missing in the mkdir array!
    mkdir => [qw(
        logs
        pid
        app/local
        app/root/cache
        app/root/files
        app/root/static/_css
        app/root/static/_js
    )],
};

### BETTER:
Perl_Modules '/path/to/website/MyApp/local' => {
    perl => Perlbrew->perl,
    cpanm => Perlbrew->cpanm,
    
    # optional: a CPAN mirror to use
    mirror => '',
    
    # --installdeps from a distribution dir
    installdeps => '/path/to/website/MyApp',
    
    # manually picked modules or tarballs
    install => [
        'Package::Name',
        'Package-Name-0.07.tar.gz',
        Resource('modules/my_package'),
    ],
};

File '/path/to/website/MyApp/static/_css/site.css' => {
    content => Execute('/path/to/binary'),
},

Execute db_migration => {
    path => '/path/to/website/MyApp/script/db_migration.pl',
};

Service plack_server => {
    user => 'sites',
    runlevel => [2,3],
    copy => Resource('...'),
};

Service thumbnail_hotfolder => {
    user => 'sites',
    runlevel => [2,3],
    copy => Resource('...'),
};


__END__

# ----- general parameter syntax:

Keyword;                    # usually dies, name is required

Keyword 'name';
Keyword name => ( ... );    # maybe unclever
Keyword name => { ... };
Keyword { ... };

Keyword();
Keyword('name');
Keyword(name => { ... });
Keyword({ ... });
